### PR TITLE
Concurrent query improvement for imodel.js 3.0

### PR DESCRIPTION
**Pull request by khanaffan**
[Corresponding native PR](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodel02/pullrequest/197543).

### CODE REVIEW

**Review by pmconne — CHANGES_REQUESTED** [review]
`rush extract-api/lint/cover/change`
https://dev.azure.com/bentleycs/iModelTechnologies/_wiki/wikis/iModelTechnologies.wiki/8690/iModel.js-Node-add-on-development

**Review by pmconne — CHANGES_REQUESTED** [review]
Please thoroughly document the changes in NextVersion.md. Include examples of how to upgrade existing code to new APIs. Add to the per-package tables rows indicating which APIs have been completely removed and which have been replaced.

**Review by pmconne — CHANGES_REQUESTED** [review]
Please merge imodel02 into your branch and run `rushchange`.

**Comment by pmconne** [comment]
Please merge imodel02 into your branch and run rushchange.

**Comment by pmconne** [comment]
@grigasp `rush cover` produced 4 failures in itwin/presentatinon-testing and now appears hung in itwin/appui-react. Is this reproducible on master?
```
==[ @itwin/presentation-testing ]================================[ 53 of 79 ]==

  7 passing (80ms)
  4 failing

  1) ContentBuilder
       createContentForAllClasses
         returns all required instances with empty records:
     TypeError: Cannot read property 'Symbol(Symbol.asyncIterator)' of undefined
      at __asyncValues (src\presentation-testing\ContentBuilder.ts:4:96)
      at ContentBuilder.createContentForClasses (src\presentation-testing\ContentBuilder.ts:57:4)
      at Context.<anonymous> (src\test\ContentBuilder.test.ts:241:23)

  2) ContentBuilder
       createContentForInstancePerClass
         test instances have ids
           returns all required instances with empty records:
     TypeError: Cannot read property 'Symbol(Symbol.asyncIterator)' of undefined
      at __asyncValues (src\presentation-testing\ContentBuilder.ts:4:96)
      at ContentBuilder.createContentForClasses (src\presentation-testing\ContentBuilder.ts:57:4)
      at Context.<anonymous> (src\test\ContentBuilder.test.ts:281:25)

  3) ContentBuilder
       createContentForInstancePerClass
         test instances have ids
           throws when id query throws an unexpected error:

      AssertionError: expected promise to be rejected with an error including 'Test error' but got 'Cannot read property \'Symbol(Symbol.asyncIterator)\' of undefined'
      + expected - actual

      -Cannot read property 'Symbol(Symbol.asyncIterator)' of undefined
      +Test error



  4) ContentBuilder
       createContentForInstancePerClass
         test instances have no ids
           returns an empty list:
     TypeError: Cannot read property 'Symbol(Symbol.asyncIterator)' of undefined
      at __asyncValues (src\presentation-testing\ContentBuilder.ts:4:96)
      at ContentBuilder.createContentForClasses (src\presentation-testing\ContentBuilder.ts:57:4)
      at Context.<anonymous> (src\test\ContentBuilder.test.ts:325:25)



npm ERR! Test failed.  See above for more details.
ERROR: Coverage for lines (90.76%) does not meet global threshold (100%)
ERROR: Coverage for functions (94.73%) does not meet global threshold (100%)
ERROR: Coverage for branches (90.47%) does not meet global threshold (100%)
ERROR: Coverage for statements (88.88%) does not meet global threshold (100%)
Returned error code: 1
"@itwin/presentation-testing" failed to build.
.....
==[ @itwin/core-backend ]========================================[ 78 of 79 ]==
WARNING: 2 tests skipped in @itwin/core-backend
"@itwin/core-backend" completed with warnings in 3 minutes 28.0 seconds.

==[ @itwin/appui-react ]=========================================[ 79 of 79 ]==
!devSeverity || (*devSeverity) < LOG_ERROR d:\i2\s\imodel02\iModelCore\ECPresentation\Source\Diagnostics.cpp 242
```

**Comment by pmconne** [comment]
@grigasp @saskliutas @roluk Here's the tail end of the ui-framework `npm run cover` output where it hangs. Please confirm whether this is a problem specific to @khanaffan's branch or something that slipped through onto the imodel02 branch.
```
  ModelsTree
    #unit
      <ModelsTree />
        √ should match snapshot
        √ renders nodes as unchecked when they're not displayed
        √ renders nodes as checked when they're displayed
        √ re-renders nodes on `onVisibilityChange` event
        √ re-renders nodes without checkboxes if visibility handler does not exist
        √ calls visibility handler's `changeVisibility` on node checkbox state changes to 'checked'
        selection
          √ adds node to unified selection
          √ adds element node to unified selection according to `selectionPredicate`
          √ adds multiple model nodes to unified selection according to `selectionPredicate`
          √ adds subject node to unified selection according to `selectionPredicate`
          √ adds node without extendedData to unified selection according to `selectionPredicate`
          √ adds element class grouping node to unified selection according to `selectionPredicate`
          √ does not add category node to unified selection according to `selectionPredicate`
        filtering
          √ filters nodes
          √ invokes onFilterApplied callback
          √ filters nodes by element IDs
    #integration
!devSeverity || (*devSeverity) < LOG_ERROR d:\i2\s\imodel02\iModelCore\ECPresentation\Source\Diagnostics.cpp 242      1) shows correct hierarchy
      2) shows correct hierarchy with class grouping
!devSeverity || (*devSeverity) < LOG_ERROR d:\i2\s\imodel02\iModelCore\ECPresentation\Source\Diagnostics.cpp 242      3) renders component with real data and no active viewport
```

**Review by pmconne — CHANGES_REQUESTED** [review]
presentation/testing has 4 failing tests. These tests all pass using an addon built from master and the imodel02 branch.

appui-react tests hang while running. Using an addon built from master and the imodel02 branch, they do not hang, but there are 49 failures.

File: common/api/core-common.api.md [hunk 1]
```diff
@@ -6161,52 +6360,123 @@ export namespace Quantization {
...
     // (undocumented)
-    PostError = 6,
+    toJsRow(): any;
     // (undocumented)
-    Timeout = 4
+    toRow(): any;
 }
 
+// @beta (undocumented)
+export type QueryValueType = any;
```
Comment (ramanujam-raman) [inline common/api/core-common.api.md:6478]: Should this just be a union of all possible types? Perhaps use a string indexer for structs, with the value being a union again?

**Comment by saskliutas** [comment]
> @grigasp @saskliutas @roluk Here's the tail end of the ui-framework `npm run cover` output where it hangs. Please confirm whether this (and the above presentation-testing failures) is a problem specific to @khanaffan's branch or something that slipped through onto the imodel02 branch.
> 
> ```
>   ModelsTree
>     #unit
>       <ModelsTree />
>         √ should match snapshot
>         √ renders nodes as unchecked when they're not displayed
>         √ renders nodes as checked when they're displayed
>         √ re-renders nodes on `onVisibilityChange` event
>         √ re-renders nodes without checkboxes if visibility handler does not exist
>         √ calls visibility handler's `changeVisibility` on node checkbox state changes to 'checked'
>         selection
>           √ adds node to unified selection
>           √ adds element node to unified selection according to `selectionPredicate`
>           √ adds multiple model nodes to unified selection according to `selectionPredicate`
>           √ adds subject node to unified selection according to `selectionPredicate`
>           √ adds node without extendedData to unified selection according to `selectionPredicate`
>           √ adds element class grouping node to unified selection according to `selectionPredicate`
>           √ does not add category node to unified selection according to `selectionPredicate`
>         filtering
>           √ filters nodes
>           √ invokes onFilterApplied callback
>           √ filters nodes by element IDs
>     #integration
> !devSeverity || (*devSeverity) < LOG_ERROR d:\i2\s\imodel02\iModelCore\ECPresentation\Source\Diagnostics.cpp 242      1) shows correct hierarchy
>       2) shows correct hierarchy with class grouping
> !devSeverity || (*devSeverity) < LOG_ERROR d:\i2\s\imodel02\iModelCore\ECPresentation\Source\Diagnostics.cpp 242      3) renders component with real data and no active viewport
> ```

Not related to @khanaffan changes. Working on a fix.

File: common/api/core-common.api.md (hunk 1 above)
Comment (khanaffan) [reply common/api/core-common.api.md:6478]: I am testing this. I ran rush cover and it fail for some other reasons for some ui component  test. I am now doing clean build to see if i can reproduce what you seeing.

==[ @itwin/imodel-components-react ]=============================[ 44 of 79 ]==

D:\bsw\m0\imodeljs\ui\imodel-components\lib\test\setup.js:24
    this.x = x ?? 0;

**Comment by khanaffan** [comment]
I am testing this. I ran rush cover and it fail for some other reasons for some ui component  test. I am now doing clean build to see if i can reproduce what you seeing.

==[ @itwin/imodel-components-react ]=============================[ 44 of 79 ]==

D:\bsw\m0\imodeljs\ui\imodel-components\lib\test\setup.js:24
    this.x = x ?? 0;

**Comment by aruniverse** [comment]
> I am testing this. I ran rush cover and it fail for some other reasons for some ui component test. I am now doing clean build to see if i can reproduce what you seeing.
> 
> ==[ @itwin/imodel-components-react ]=============================[ 44 of 79 ]==
> 
> D:\bsw\m0\imodeljs\ui\imodel-components\lib\test\setup.js:24 this.x = x ?? 0;

You need to change your node-ver to the latest 14 LTS

**Comment by pmconne** [comment]
> You need to change your node-ver to the latest 14 LTS to be able to use Nullish Coalescing, see [developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator#browser_compatibility](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator#browser_compatibility)

NextVersion.md for 3.0 claims the minimum supported node version is 12.22.0. None of our current CI jobs appear to build with node 12. @calebmshafer is our node 12 support a lie?

**Comment by pmconne** [comment]
I am cleaning up the docs. I think ECSqlRowFormat.md should be updated to include info about the new row formats, or at the very least make clear that it no longer describes the default format.

**Comment by pmconne** [comment]
`QueryOptions` can't be marked `beta` - it is used by public methods like `query`.

File: core/common/src/ConcurrentQuery.ts [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+    const name = String(indexOrName);
+    Object.defineProperty(this._args, name, {
+      enumerable: true, value: {
+        type: QueryParamType.Id,
+        value: val,
+      },
+    });
+    return this;
+  }
+  public bindIdSet(indexOrName: string | number, val: Id64Set) {
```
Comment (pmconne) [inline core/common/src/ConcurrentQuery.ts:209]: This should take an `OrderedId64Iterable` and call `CompressedId64Set.compressIds`.

**Review by pmconne — CHANGES_REQUESTED** [review]

File: core/backend/src/ECDb.ts [hunk 3]
```diff
@@ -283,244 +281,91 @@ export class ECDb implements IDisposable {
...
-            setTimeout(() => { poll(); }, IModelHost.configuration!.concurrentQuery.pollInterval);
-          else if (pollResult.status === IModelJsNative.ConcurrentQuery.PollStatus.Cancelled)
-            resolve({ status: QueryResponseStatus.Cancelled, rows: [pollResult.result] });
-          else
-            resolve({ status: QueryResponseStatus.Error, rows: [pollResult.result] });
-        }
-      };
-      setTimeout(() => { poll(); }, IModelHost.configuration!.concurrentQuery.pollInterval);
-    });
+  public async * query(ecsql: string, params?: QueryBinder, rowFormat?: QueryRowFormat, config?: QueryOptions): AsyncIterableIterator<any> {
```
Comment (pmconne) [inline core/backend/src/ECDb.ts:318]: Here and elsewhere, replace `rowFormat?: QueryRowFormat` with `rowFormat: QueryRowFormat = QueryRowFormat.UseArrayIndexes` to make it clear at the call site what the default is.
`QueryRowFormat.Default` is unnecessary.

File: core/common/src/ConcurrentQuery.ts [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+  /** Each row is an object in which each non-null column value can be accessed by a [remapped property name]($docs/learning/ECSqlRowFormat.md).
+   * This format is backwards-compatible with the format produced by iModel.js version 2.x. Null values are omitted.
+   */
+  UseJsPropertyNames,
+  /** Each row is an array of values accessed by an index corresponding to the property's position in the ECSql SELECT statement.
+   * Null values are included if they are followed by a non-null column, but trailing null values at the end of the array are omitted.
+   */
+  UseArrayIndexes,
+  /** The default format if none is specified by the caller. */
+  Default = UseArrayIndexes,
```
Comment (pmconne) [inline core/common/src/ConcurrentQuery.ts:28]: `Default` is unnecessary - the default should be specified as default argument in function signatures.

File: docs/changehistory/NextVersion.md [hunk 5]
```diff
@@ -1155,3 +1155,115 @@ were removed from the `@itwin/core-backend` package and moved to a new package,
...
+
+[ECSqlReader]($common) is a flexible API that is used internally by methods like [IModelDb.query]($backend). It is also useful when you want the result to include metadata. For example:
+
+```ts
+  const params = new QueryBinder().bindString("name", "CompositeUnitRefersToUnit");
+  const reader = conn.createQueryReader("SELECT ECInstanceId, Name FROM meta.ECClassDef WHERE Name=:name");
+
+  // The first time step is called, the metadata is populated.
+  while (await reader.step()) {
+    if (needMetaData)
```
Comment (pmconne) [inline docs/changehistory/NextVersion.md:1247]: `needMetaData` and `setupColumns` are not defined. Should make more clear that at any time you can call getMetaData on the reader, or on any row.

**Review by pmconne — CHANGES_REQUESTED** [review]

File: common/api/core-common.api.md (hunk 1 above)
Comment (pmconne) [reply common/api/core-common.api.md:6478]: @kabentley talked about that a little while ago...the union would be effectively equivalent to `any` anyway.

File: presentation/testing/src/presentation-testing/ContentBuilder.ts [hunk 6]
```diff
@@ -3,17 +3,18 @@
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
 
+import { PropertyRecord } from "@itwin/appui-abstract";
```
Comment (grigasp) [inline presentation/testing/src/presentation-testing/ContentBuilder.ts:6]: this needs to be moved next to other imports

**Review by pmconne — APPROVED** [review]

File: presentation/testing/src/presentation-testing/ContentBuilder.ts (hunk 6 above)
Comment (pmconne) [reply presentation/testing/src/presentation-testing/ContentBuilder.ts:6]: moved.

**Pull request merged**